File: solutions/go/src/playground/algorithm/single_number_2.go
package algorithm

import "strconv"

func singleNumber2(nums []int) int {
	result, bits := 0, make([]int, strconv.IntSize)

	for _, n := range nums {
		for j := uint(0); j < strconv.IntSize; j++ {
			bits[j] += (n >> j) & 1
		}
	}

	for i, b := range bits {
		if b%3 > 0 {
			result += 1 << uint(i)
		}
	}

	return result
}

File: solutions/java/src/main/java/playground/algorithm/SingleNumber2.java
package playground.algorithm;

public class SingleNumber2 {

    public int singleNumber(int[] nums) {
        int result = 0;
        int[] bits = new int[32];

        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < bits.length; j++) {
                bits[j] += (nums[i] >> j) & 1;
            }
        }

        for (int i = 0; i < bits.length; i++) {
            if (bits[i] % 3 > 0) {
                result += 1 << i;
            }
        }

        return result;
    }
}
